### PR TITLE
[Apache v2] Implement find_ancestors

### PR DESCRIPTION
**Pull request by joohoi**
This PR implements find_ancestors that's needed by the Augeas ParserNode implementation to traverse ancestors where possible. While there's quite a few lines changed, majority of it is just moving few helper functions from `AugeasBlockNode` to `AugeasParserNode`, as all the node types need the `find_ancestors` capability.

Note: due to the nature of Augeas, we are not able to trace back the complete ancestry tree, for example in a VirtualHost file that's included from the root configuration, the Augeas path would be `/path/to/vhost.conf/Ancestor/VirtualHost`, leaving ancestor blocks that were defined in the main configuration invisible. This however isn't a regression from our current implementation in `master`, but apacheconfig will be able to do better here.

### CODE REVIEW

File: certbot-apache/certbot_apache/augeasparser.py [hunk 1]
```diff
@@ -100,6 +99,65 @@ def __init__(self, **kwargs):
     def save(self, msg):
         self.parser.save(msg)
 
+    def find_ancestors(self, name):
+        """
+        Searches ancestor BlockNodes with a given name.
```
Comment (ohemorange) [inline certbot-apache/certbot_apache/augeasparser.py]:
```suggestion
        Searches for ancestor BlockNodes with a given name.
```

File: certbot-apache/certbot_apache/augeasparser.py [hunk 2]
```diff
@@ -100,6 +99,65 @@ def __init__(self, **kwargs):
...
+        AugeasParserNode.find_ancestors and AugeasBlockNode.
+        and AugeasBlockNode.find_blocks
+
+        """
+
+        name = self._aug_get_name(path)
+        metadata = {"augeasparser": self.parser, "augeaspath": path}
+
+        # Because of the dynamic nature, and the fact that we're not populating
+        # the complete ParserNode tree, we use the search parent as ancestor
```
Comment (ohemorange) [inline certbot-apache/certbot_apache/augeasparser.py]: I thought we removed this comment, but no time like the present to do so.

File: certbot-apache/certbot_apache/augeasparser.py [hunk 3]
```diff
@@ -100,6 +99,65 @@ def __init__(self, **kwargs):
...
+        :returns: List of matching ancestor nodes.
+        :rtype: list of AugeasBlockNode
+        """
+
+        ancestors = []
+
+        parent = self.metadata["augeaspath"]
+        while True:
+            # Get the path of ancestor node
+            parent = parent.rpartition("/")[0]
```
Comment (ohemorange) [inline certbot-apache/certbot_apache/augeasparser.py]: does augeas always specifically use "/" in an api-defined way, or should we actually be looking for something like `pathsep`?

Comment (joohoi) [reply certbot-apache/certbot_apache/augeasparser.py]: Yeah, Augeas always uses `/` in values returned from the searches as they always point to an absolute location.

**Pull request merged**